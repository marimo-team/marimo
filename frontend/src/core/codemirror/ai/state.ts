/* Copyright 2024 Marimo. All rights reserved. */

import { singleFacet } from "../facet";

interface ContextCallbacks {
  addAttachment?: (attachment: File) => void;
}

/**
 * State for completion callbacks
 */
export const contextCallbacks = singleFacet<ContextCallbacks>();
