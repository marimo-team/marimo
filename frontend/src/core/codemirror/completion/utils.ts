import type { CompletionSource } from "@codemirror/autocomplete";

/**
 * Condition CompletionSource
 */
export function conditionalCompletion(opts: {
  completion: CompletionSource;
  predicate: () => boolean;
}): CompletionSource {
  return (ctx) => {
    if (!opts.predicate()) {
      return null;
    }
    return opts.completion(ctx);
  };
}
