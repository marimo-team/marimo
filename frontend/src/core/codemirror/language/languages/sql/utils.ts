/* Copyright 2024 Marimo. All rights reserved. */

import type { SQLDialect } from "@codemirror/lang-sql";
import {
  Cassandra,
  MariaSQL,
  MSSQL,
  MySQL,
  PLSQL,
  PostgreSQL,
  SQLite,
} from "@codemirror/lang-sql";
import { DuckDBDialect } from "@marimo-team/codemirror-sql/dialects";
import type { DataSourceConnection } from "@/core/kernel/messages";

export function guessDialect(
  connection: DataSourceConnection,
): SQLDialect | undefined {
  switch (connection.dialect) {
    case "postgresql":
    case "postgres":
      return PostgreSQL;
    case "mysql":
      return MySQL;
    case "sqlite":
      return SQLite;
    case "mssql":
    case "sqlserver":
      return MSSQL;
    case "duckdb":
      return DuckDBDialect;
    case "mariadb":
      return MariaSQL;
    case "cassandra":
      return Cassandra;
    case "oracledb":
    case "oracle":
      return PLSQL;
    default:
      return undefined;
  }
}
