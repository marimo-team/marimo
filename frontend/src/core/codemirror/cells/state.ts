/* Copyright 2024 Marimo. All rights reserved. */
import type { CellId } from "@/core/cells/ids";
import { Facet } from "@codemirror/state";
import type { CellActions } from "@/core/cells/cells";

export interface CodemirrorCellActions extends CellActions {
  toggleHideCode: () => boolean;
  aiCellCompletion: () => boolean;
  createManyBelow: (content: string[]) => void;
  onRun: () => void;
  deleteCell: () => void;
  afterToggleMarkdown: () => void;
}

/**
 * State for cell actions
 */
export const cellActionsState = Facet.define<
  CodemirrorCellActions,
  CodemirrorCellActions
>({
  combine: (values) => values[0],
});

/**
 * State for cell id
 */
export const cellIdState = Facet.define<CellId, CellId>({
  combine: (values) => values[0],
});
