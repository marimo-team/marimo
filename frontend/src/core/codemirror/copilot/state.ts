/* Copyright 2024 Marimo. All rights reserved. */
import { getUserConfig, userConfigAtom } from "@/core/config/config";
import { waitFor } from "@/core/state/jotai";
import { atomWithStorage } from "jotai/utils";

const KEY = "marimo:copilot:signedIn";

export const copilotSignedInState = atomWithStorage<boolean | null>(
  KEY,
  null,
  undefined,
  {
    getOnInit: true,
  },
);

function getIsLastSignedIn() {
  const lastSignedIn = localStorage.getItem(KEY);
  return lastSignedIn === "true";
}

export function isCopilotEnabled() {
  const copilot = getIsLastSignedIn();
  const userConfig = getUserConfig();
  return copilot && userConfig.completion.copilot;
}

export function waitForEnabledCopilot() {
  return waitFor(userConfigAtom, (value) => {
    return value.completion.copilot;
  });
}
