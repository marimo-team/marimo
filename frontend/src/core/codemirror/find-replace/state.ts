/* Copyright 2024 Marimo. All rights reserved. */
import { store } from "@/core/state/jotai";
import { EditorView } from "@codemirror/view";
import { atomWithReducer } from "jotai/utils";

interface FindReplaceState {
  findText: string;
  replaceText: string;
  caseSensitive: boolean;
  wholeWord: boolean;
  regexp: boolean;
  isOpen: boolean;
  currentView?: {
    view: EditorView;
    range: { from: number; to: number };
  };
}

type Action =
  | {
      type: "setFind";
      find: string;
    }
  | {
      type: "setReplace";
      replace: string;
    }
  | {
      type: "setCaseSensitive";
      caseSensitive: boolean;
    }
  | {
      type: "setWholeWord";
      wholeWord: boolean;
    }
  | {
      type: "setRegex";
      regexp: boolean;
    }
  | {
      type: "setIsOpen";
      isOpen: boolean;
    }
  | {
      type: "setCurrentView";
      view: EditorView;
      range: { from: number; to: number };
    }
  | {
      type: "clearCurrentView";
    };

export const findReplaceAtom = atomWithReducer<FindReplaceState, Action>(
  {
    findText: "",
    replaceText: "",
    caseSensitive: false,
    wholeWord: false,
    regexp: false,
    isOpen: false,
    currentView: undefined,
  },
  (state, action) => {
    if (action === undefined) {
      return state;
    }

    switch (action.type) {
      case "setFind":
        return { ...state, findText: action.find };
      case "setReplace":
        return { ...state, replaceText: action.replace };
      case "setCaseSensitive":
        return { ...state, caseSensitive: action.caseSensitive };
      case "setWholeWord":
        return { ...state, wholeWord: action.wholeWord };
      case "setRegex":
        return { ...state, regexp: action.regexp };
      case "setIsOpen":
        return { ...state, isOpen: action.isOpen };
      case "setCurrentView":
        return {
          ...state,
          currentView: { view: action.view, range: action.range },
        };
      case "clearCurrentView":
        return { ...state, currentView: undefined };
    }
  },
);

export function openFindReplacePanel(initialView?: EditorView): boolean {
  // If any radix dialog is open, don't open the find/replace panel
  // they have role="dialog" and data-state="open"
  const element = document.querySelector('[role="dialog"][data-state="open"]');
  if (element) {
    return false;
  }

  if (initialView) {
    // Set the selected text and focus
    const selection = initialView.state.selection.main;
    const query = initialView.state.sliceDoc(selection.from, selection.to);
    if (query) {
      store.set(findReplaceAtom, {
        type: "setFind",
        find: query,
      });
      store.set(findReplaceAtom, {
        type: "setCurrentView",
        view: initialView,
        range: { from: selection.from, to: selection.to },
      });
    } else {
      // If there is no selection, just set the current view
      store.set(findReplaceAtom, {
        type: "setCurrentView",
        view: initialView,
        range: { from: 0, to: 0 },
      });
    }
  }

  // HACK: If open, close it and re-open it to gain focus
  const isOpen = store.get(findReplaceAtom).isOpen;
  if (isOpen) {
    store.set(findReplaceAtom, {
      type: "setIsOpen",
      isOpen: false,
    });
    requestAnimationFrame(() => {
      // Open the panel
      store.set(findReplaceAtom, {
        type: "setIsOpen",
        isOpen: true,
      });
    });
    return true;
  }

  // Open the panel
  store.set(findReplaceAtom, {
    type: "setIsOpen",
    isOpen: true,
  });
  return true;
}

export function closeFindReplacePanel(): boolean {
  const isOpen = store.get(findReplaceAtom).isOpen;
  // If the panel is already closed, return false.
  if (!isOpen) {
    return false;
  }

  store.set(findReplaceAtom, {
    type: "setIsOpen",
    isOpen: false,
  });
  return true;
}
