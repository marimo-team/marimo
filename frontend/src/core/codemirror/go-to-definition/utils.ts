/* Copyright 2024 Marimo. All rights reserved. */
import type { EditorState } from "@codemirror/state";
import { closeHoverTooltips, type EditorView } from "@codemirror/view";
import { getPositionAtWordBounds } from "../completion/hints";
import type { VariableName, Variables } from "../../variables/types";
import { store } from "../../state/jotai";
import { notebookAtom } from "../../cells/cells";
import { variablesAtom } from "../../variables/state";
import type { CellId } from "@/core/cells/ids";
import { goToVariableDefinition, goToLine } from "./commands";
import { closeCompletion } from "@codemirror/autocomplete";

/**
 * Get the word under the cursor.
 */
function getWordUnderCursor(state: EditorState) {
  const { from, to } = state.selection.main;
  if (from === to) {
    const { startToken, endToken } = getPositionAtWordBounds(state.doc, from);
    return state.doc.sliceString(startToken, endToken);
  }

  return state.doc.sliceString(from, to);
}

/**
 * Get the cell id of the definition of the given variable.
 */
function getCellIdOfDefinition(
  variables: Variables,
  variableName: string,
): CellId | null {
  if (!variableName) {
    return null;
  }
  const variable = variables[variableName as VariableName];
  if (!variable || variable.declaredBy.length === 0) {
    return null;
  }
  return variable.declaredBy[0];
}

function isPrivateVariable(variableName: string) {
  return variableName.startsWith("_");
}

/**
 * Go to the definition of the variable under the cursor.
 * @param view The editor view at which the command was invoked.
 */
export function goToDefinitionAtCursorPosition(view: EditorView): boolean {
  const { state } = view;
  const variableName = getWordUnderCursor(state);
  if (!variableName) {
    return false;
  }
  // Close popovers/tooltips
  closeCompletion(view);
  view.dispatch({ effects: closeHoverTooltips });

  return goToDefinition(view, variableName);
}

/**
 * Go to the definition of the variable under the cursor.
 * @param view The editor view at which the command was invoked.
 */
export function goToDefinition(
  view: EditorView,
  variableName: string,
): boolean {
  // The variable may exist in another cell
  const editorWithVariable = getEditorForVariable(view, variableName);
  if (!editorWithVariable) {
    return false;
  }
  return goToVariableDefinition(editorWithVariable, variableName);
}

/**
 * Go to the given line number in the cell with the given ID.
 * @param cellId The ID of the cell to go to.
 * @param line The line number to go to.
 */
export function goToCellLine(cellId: CellId, lineNumber: number): boolean {
  const view = getEditorForCell(cellId);
  if (!view) {
    return false;
  }
  return goToLine(view, lineNumber);
}

/**
 * @param editor The editor view at which the command was invoked.
 * @param variableName  The name of the variable to go to.
 */
function getEditorForVariable(
  editor: EditorView,
  variableName: string,
): EditorView | null {
  // If it's a private variable, we only want to go to the
  // definition if it's in the same cell
  if (isPrivateVariable(variableName)) {
    return editor;
  }

  const variables = store.get(variablesAtom);

  const cellId = getCellIdOfDefinition(variables, variableName);
  if (cellId) {
    return getEditorForCell(cellId);
  }

  return null;
}

/**
 * Go to the given line number in the editor view.
 * @param view The editor view to go to.
 * @param line The line number to go to.
 */
function getEditorForCell(cellId: CellId): EditorView | null {
  const notebookState = store.get(notebookAtom);
  return notebookState.cellHandles[cellId].current?.editorView ?? null;
}
