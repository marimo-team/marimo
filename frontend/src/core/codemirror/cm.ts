/* Copyright 2024 Marimo. All rights reserved. */
import {
  acceptCompletion,
  closeBrackets,
  closeBracketsKeymap,
} from "@codemirror/autocomplete";
import {
  history,
  historyKeymap,
  indentWithTab,
  indentMore,
} from "@codemirror/commands";
import {
  bracketMatching,
  defaultHighlightStyle,
  foldGutter,
  foldKeymap,
  indentOnInput,
  indentUnit,
  syntaxHighlighting,
} from "@codemirror/language";
import { lintKeymap } from "@codemirror/lint";
import {
  drawSelection,
  dropCursor,
  highlightActiveLine,
  highlightActiveLineGutter,
  highlightSpecialChars,
  lineNumbers,
  keymap,
  rectangularSelection,
  tooltips,
  EditorView,
} from "@codemirror/view";

import { EditorState, Extension, Prec } from "@codemirror/state";
import { oneDark } from "@codemirror/theme-one-dark";

import { CompletionConfig, KeymapConfig } from "../config/config-schema";
import { Theme } from "../../theme/useTheme";

import { findReplaceBundle } from "./find-replace/extension";
import {
  CodeCallbacks,
  MovementCallbacks,
  cellCodeEditingBundle,
  cellMovementBundle,
} from "./cells/extensions";
import { CellId } from "../cells/ids";
import { keymapBundle } from "./keymaps/keymaps";
import {
  scrollActiveLineIntoView,
  smartPlaceholderExtension,
} from "./extensions";
import { copilotBundle } from "./copilot/extension";
import { hintTooltip } from "./completion/hints";
import { adaptiveLanguageConfiguration } from "./language/extension";
import { historyCompartment } from "./editing/extensions";

export interface CodeMirrorSetupOpts {
  cellId: CellId;
  showPlaceholder: boolean;
  cellMovementCallbacks: MovementCallbacks;
  cellCodeCallbacks: CodeCallbacks;
  completionConfig: CompletionConfig;
  keymapConfig: KeymapConfig;
  theme: Theme;
}

/**
 * Setup CodeMirror for a cell
 */
export const setupCodeMirror = ({
  cellId,
  showPlaceholder,
  cellMovementCallbacks,
  cellCodeCallbacks,
  completionConfig,
  keymapConfig,
  theme,
}: CodeMirrorSetupOpts): Extension[] => {
  return [
    // Editor keymaps (vim or defaults) based on user config
    keymapBundle(keymapConfig, cellMovementCallbacks),
    // Cell editing
    cellMovementBundle(cellId, cellMovementCallbacks),
    cellCodeEditingBundle(cellId, cellCodeCallbacks),
    // Comes last so that it can be overridden
    basicBundle(completionConfig, theme),
    showPlaceholder
      ? Prec.highest(smartPlaceholderExtension("import marimo as mo"))
      : [],
  ];
};

// Based on codemirror's basicSetup extension
export const basicBundle = (
  completionConfig: CompletionConfig,
  theme: Theme,
): Extension[] => {
  return [
    ///// View
    EditorView.lineWrapping,
    drawSelection(),
    dropCursor(),
    highlightActiveLine(),
    highlightActiveLineGutter(),
    highlightSpecialChars(),
    lineNumbers(),
    rectangularSelection(),
    tooltips({
      // Having fixed position prevents tooltips from being repositioned
      // and bouncing distractingly
      position: "fixed",
      // This the z-index multiple tooltips being stacked
      // For example, if we have a hover tooltip and a completion tooltip
      parent: document.querySelector<HTMLElement>("#App") ?? undefined,
    }),
    scrollActiveLineIntoView(),
    theme === "dark" ? oneDark : [],

    hintTooltip(),
    copilotBundle(),
    foldGutter(),
    closeBrackets(),
    // to avoid clash with charDeleteBackward keymap
    Prec.high(keymap.of(closeBracketsKeymap)),
    bracketMatching(),
    indentOnInput(),
    indentUnit.of("    "),
    syntaxHighlighting(defaultHighlightStyle, { fallback: true }),
    keymap.of([...foldKeymap, ...lintKeymap]),

    ///// Language Support
    adaptiveLanguageConfiguration(completionConfig),

    ///// Editing
    historyCompartment.of(history()),
    EditorState.allowMultipleSelections.of(true),
    findReplaceBundle(),
    keymap.of([
      {
        key: "Tab",
        run: (cm) => {
          return acceptCompletion(cm) || indentMore(cm);
        },
        preventDefault: true,
      },
    ]),
    keymap.of([...historyKeymap, indentWithTab]),
  ];
};

// Use the default keymap for completion
export { completionKeymap as autocompletionKeymap } from "@codemirror/autocomplete";
