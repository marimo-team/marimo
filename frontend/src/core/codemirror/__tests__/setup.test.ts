/* Copyright 2024 Marimo. All rights reserved. */
import { describe, expect, test, vi } from "vitest";
import { CodeMirrorSetupOpts, setupCodeMirror } from "../cm";
import { EditorState, Extension } from "@codemirror/state";
import { keymap } from "@codemirror/view";
import { CellId } from "@/core/cells/ids";
import { Objects } from "@/utils/objects";

vi.mock("@/core/config/config", () => ({
  parseAppConfig: () => ({}),
  parseUserConfig: () => ({}),
}));

function namedFunction(name: string) {
  const fn = () => false;
  Object.defineProperty(fn, "name", { value: name });
  return fn;
}

function setup(config: Partial<CodeMirrorSetupOpts> = {}): Extension[] {
  return setupCodeMirror({
    cellId: "0" as CellId,
    showPlaceholder: false,
    cellMovementCallbacks: {
      onRun: namedFunction("onRun"),
      deleteCell: namedFunction("deleteCell"),
      createAbove: namedFunction("createAbove"),
      createBelow: namedFunction("createBelow"),
      moveUp: namedFunction("moveUp"),
      moveDown: namedFunction("moveDown"),
      focusUp: namedFunction("focusUp"),
      focusDown: namedFunction("focusDown"),
      sendToTop: namedFunction("sendToTop"),
      sendToBottom: namedFunction("sendToBottom"),
      moveToNextCell: namedFunction("moveToNextCell"),
      toggleHideCode: namedFunction("toggleHideCode"),
    },
    cellCodeCallbacks: {
      updateCellCode: namedFunction("updateCellCode"),
    },
    completionConfig: {
      activate_on_typing: false,
      copilot: false,
    },
    keymapConfig: {
      preset: "default",
    },
    theme: "light",
    ...config,
  });
}

function prettyPrintKeymaps(state: EditorState) {
  const keymaps = state.facet(keymap).flat();
  const prettyKeymaps = keymaps.map((keymap) => {
    const { key, run, any, shift, ...rest } = keymap;
    return {
      key: key?.toString(),
      ...(any ? { any: any.name || "<no name>" } : {}),
      ...(run ? { run: run.name || "<no name>" } : {}),
      ...(shift ? { shift: shift.name || "<no name>" } : {}),
      ...rest,
    };
  });
  return prettyKeymaps;
}

function getDuplicateKeymaps(state: EditorState) {
  const prettyKeymaps = prettyPrintKeymaps(state);
  const groupBy = Objects.groupBy(
    prettyKeymaps,
    (keymap) => keymap.key,
    (keymap) => keymap,
  );
  const duplicates = Objects.fromEntries(
    Object.entries(groupBy).filter(([key, value]) => value.length > 1),
  );
  return duplicates;
}

describe("snapshot all duplicate keymaps", () => {
  // This test just ensures we are not accidentally overlapping keymaps
  // without handling it (precedence or otherwise).

  test("default keymaps", () => {
    const extensions = setup();
    const duplicates = getDuplicateKeymaps(
      EditorState.create({ extensions: extensions }),
    );
    // Total duplicates:
    // if this changes, please make sure to validate they are not conflicting
    expect(Object.values(duplicates).flat().length).toMatchInlineSnapshot(`20`);
    expect(duplicates).toMatchSnapshot();
  });

  test("vim keymaps", () => {
    const extensions = setup({
      keymapConfig: { preset: "vim" },
    });
    const duplicates = getDuplicateKeymaps(
      EditorState.create({ extensions: extensions }),
    );
    // Total duplicates:
    // if this changes, please make sure to validate they are not conflicting
    expect(Object.values(duplicates).flat().length).toMatchInlineSnapshot(`19`);
    expect(duplicates).toMatchSnapshot();
  });
});

test("placeholder adds another extension", () => {
  const withPlaceholder = setup({ showPlaceholder: true }).flat();
  const withoutPlaceholder = setup({ showPlaceholder: false }).flat();
  expect(withPlaceholder.length - 1).toBe(withoutPlaceholder.length);
});
