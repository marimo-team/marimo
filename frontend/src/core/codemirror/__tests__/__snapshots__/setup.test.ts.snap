// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`snapshot all duplicate keymaps > default keymaps 2`] = `
{
  "ArrowDown": [
    {
      "key": "ArrowDown",
      "preventDefault": true,
      "run": "run",
      "stopPropagation": true,
    },
    {
      "key": "ArrowDown",
      "run": "<no name>",
    },
    {
      "key": "ArrowDown",
      "preventDefault": true,
      "run": "cursorLineDown",
      "shift": "selectLineDown",
    },
  ],
  "ArrowUp": [
    {
      "key": "ArrowUp",
      "preventDefault": true,
      "run": "run",
      "stopPropagation": true,
    },
    {
      "key": "ArrowUp",
      "run": "<no name>",
    },
    {
      "key": "ArrowUp",
      "preventDefault": true,
      "run": "cursorLineUp",
      "shift": "selectLineUp",
    },
  ],
  "Backspace": [
    {
      "key": "Backspace",
      "run": "deleteBracketPair",
    },
    {
      "key": "Backspace",
      "run": "deleteCharBackward",
      "shift": "deleteCharBackward",
    },
  ],
  "Enter": [
    {
      "key": "Enter",
      "run": "acceptCompletion",
    },
    {
      "key": "Enter",
      "run": "<no name>",
    },
  ],
  "Escape": [
    {
      "key": "Escape",
      "run": "closeCompletion",
    },
    {
      "key": "Escape",
      "run": "simplifySelection",
    },
    {
      "key": "Escape",
      "preventDefault": true,
      "run": "run",
    },
    {
      "key": "Escape",
      "preventDefault": false,
      "run": "closeFindReplacePanel",
    },
  ],
  "PageDown": [
    {
      "key": "PageDown",
      "run": "<no name>",
    },
    {
      "key": "PageDown",
      "run": "cursorPageDown",
      "shift": "selectPageDown",
    },
  ],
  "PageUp": [
    {
      "key": "PageUp",
      "run": "<no name>",
    },
    {
      "key": "PageUp",
      "run": "cursorPageUp",
      "shift": "selectPageUp",
    },
  ],
  "Tab": [
    {
      "key": "Tab",
      "preventDefault": true,
      "run": "run",
    },
    {
      "key": "Tab",
      "run": "indentMore",
      "shift": "indentLess",
    },
  ],
}
`;

exports[`snapshot all duplicate keymaps > vim keymaps 2`] = `
{
  "ArrowDown": [
    {
      "key": "ArrowDown",
      "preventDefault": true,
      "run": "run",
      "stopPropagation": true,
    },
    {
      "key": "ArrowDown",
      "run": "<no name>",
    },
    {
      "key": "ArrowDown",
      "preventDefault": true,
      "run": "cursorLineDown",
      "shift": "selectLineDown",
    },
  ],
  "ArrowUp": [
    {
      "key": "ArrowUp",
      "preventDefault": true,
      "run": "run",
      "stopPropagation": true,
    },
    {
      "key": "ArrowUp",
      "run": "<no name>",
    },
    {
      "key": "ArrowUp",
      "preventDefault": true,
      "run": "cursorLineUp",
      "shift": "selectLineUp",
    },
  ],
  "Backspace": [
    {
      "key": "Backspace",
      "run": "deleteBracketPair",
    },
    {
      "key": "Backspace",
      "run": "deleteCharBackward",
      "shift": "deleteCharBackward",
    },
  ],
  "Enter": [
    {
      "key": "Enter",
      "run": "acceptCompletion",
    },
    {
      "key": "Enter",
      "run": "<no name>",
    },
  ],
  "Escape": [
    {
      "key": "Escape",
      "run": "closeCompletion",
    },
    {
      "key": "Escape",
      "run": "simplifySelection",
    },
    {
      "key": "Escape",
      "preventDefault": false,
      "run": "closeFindReplacePanel",
    },
  ],
  "PageDown": [
    {
      "key": "PageDown",
      "run": "<no name>",
    },
    {
      "key": "PageDown",
      "run": "cursorPageDown",
      "shift": "selectPageDown",
    },
  ],
  "PageUp": [
    {
      "key": "PageUp",
      "run": "<no name>",
    },
    {
      "key": "PageUp",
      "run": "cursorPageUp",
      "shift": "selectPageUp",
    },
  ],
  "Tab": [
    {
      "key": "Tab",
      "preventDefault": true,
      "run": "run",
    },
    {
      "key": "Tab",
      "run": "indentMore",
      "shift": "indentLess",
    },
  ],
}
`;
